### PR TITLE
Put publics inside namespace.

### PR DESCRIPTION
**Pull request by chriseth**
We inherited from polygon the speciality that publics are always global even if declared inside a namespace. This PR changes that and treats publics just as any other definition. It also does some clean-up in that the degrees of the namespaces are first determined in the Display function and not only while printing. This allows declarations without degree to be used as a first declaration in a namespace.

### CODE REVIEW

File: pil-analyzer/src/statement_processor.rs [hunk 1]
```diff
@@ -515,6 +515,7 @@ where
         index: parsed::Expression,
     ) -> Vec<PILItem> {
         let id = self.counters.dispense_public_id();
+        let name = self.driver.resolve_decl(&name);
```
Comment (chriseth) [inline pil-analyzer/src/statement_processor.rs:518]: This is the main change...

**Comment by leonardoalt** [comment]
@georgwiese @Schaeff does the split_pil stuff still work after this?

File: ast/src/analyzed/display.rs [hunk 2]
```diff
@@ -31,27 +31,43 @@ impl Display for DegreeRange {
...
+            .values()
+            .filter_map(|(symbol, _)| {
+                symbol.degree.map(|degree| {
+                    let (namespace, _) = split_into_namespace_and_name(&symbol.absolute_name);
+                    (namespace, degree)
+                })
+            })
+            .unique()
+            .fold(BTreeMap::default(), |mut degrees, (namespace, degree)| {
+                let inserted = degrees.insert(namespace.clone(), degree).is_none();
```
Comment (georgwiese) [inline ast/src/analyzed/display.rs:46]:
```suggestion
                let is_new = degrees.insert(namespace.clone(), degree).is_none();
```
It's always inserted, right? Could also be inlined inside the `assert` IMO.

**Comment by leonardoalt** [comment]
@georgwiese it still happens in this branch

**Comment by georgwiese** [comment]
Can you post the optimized PIL?

**Comment by leonardoalt** [comment]
https://gist.github.com/leonardoalt/300c597a3ba41fcd13fc14ebb34e6a9c

**Review by georgwiese — APPROVED** [review]

**Pull request merged**